rollup/l1/msg_storage.go

```diff
@@ -0,0 +1,372 @@
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rlp"
+)
+
+const (
+	defaultFetchInterval   = 5 * time.Second
+	maxMsgsStored          = 2000
+	defaultFetchBlockRange = uint64(500)
+)
+
+type MsgStorageState struct {
+	StartBlockHeader *types.Header
+	EndBlockHeader   *types.Header
+}
+
+type MsgStorage struct {
+	state MsgStorageState
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	db                    ethdb.Database
+	msgs                  *common.ShrinkingMap[uint64, storedL1Message]
+	reader                *Reader
+	tracker               *Tracker
+	confirmationRule      ConfirmationRule
+	unsubscribeTracker    func()
+	newChainNotifications chan newChainNotification
+	latestFinalized       *types.Header
+
+	msgsMu sync.RWMutex
+}
+
+func NewMsgStorage(ctx context.Context, tracker *Tracker, reader *Reader, db ethdb.Database, confirmationRule ConfirmationRule, l1DeploymentBlockHeader *types.Header) (*MsgStorage, error) {
+	if tracker == nil || reader == nil {
+		return nil, fmt.Errorf("failed to create MsgStorage, reader or tracker is nil")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	msgStorage := &MsgStorage{
+		ctx:                   ctx,
+		cancel:                cancel,
+		db:                    db,
+		msgs:                  common.NewShrinkingMap[uint64, storedL1Message](maxMsgsStored),
+		reader:                reader,
+		tracker:               tracker,
+		confirmationRule:      confirmationRule,
+		newChainNotifications: make(chan newChainNotification, 10),
+		state: MsgStorageState{
+			StartBlockHeader: l1DeploymentBlockHeader,
+			EndBlockHeader:   l1DeploymentBlockHeader,
+		},
+		latestFinalized: l1DeploymentBlockHeader,
+	}
+	return msgStorage, nil
+}
+
+func (ms *MsgStorage) Start() {
+	log.Info("starting MsgStorage")
+
+	data := rawdb.ReadL1MsgStorageState(ms.db)
+	state := &MsgStorageState{}
+	if err := rlp.Decode(bytes.NewReader(data), state); err == nil {
+		ms.state = *state
+		ms.resyncToState()
+	}
+	ms.unsubscribeTracker = ms.tracker.Subscribe(ms.confirmationRule, func(isOnline bool, old, new []*types.Header) bool {
+		if isOnline {
+			ms.newChainNotifications <- newChainNotification{old, new}
+			return true
+		} else {
+			ms.latestFinalized = new[0]
+			// check if latest synced block is already finalized or later -> can become an online subscriber
+			if ms.state.EndBlockHeader.Number.Uint64() >= new[0].Number.Uint64() {
+				return true
+			}
+		}
+		return false
+	})
+	go func() {
+		fetchTicker := time.NewTicker(defaultFetchInterval)
+		defer fetchTicker.Stop()
+
+		for {
+			select {
+			case <-ms.ctx.Done():
+				return
+			default:
+			}
+			select {
+			case <-ms.ctx.Done():
+				return
+			case <-fetchTicker.C:
+				if ms.state.EndBlockHeader.Number.Uint64() < ms.latestFinalized.Number.Uint64() {
```

Comment:
here we should just use whatever we read from the channel. if it is far away we might need to have a special way of handling it. if it is close, then we can simply request all of the L1 messages.